File: LeetCode_ProblemSet/Go/Problem885.go
package main

func spiralMatrixIII(rows int, cols int, rStart int, cStart int) [][]int {
    cell_count := rows * cols
	next_direction := map[string]string{
		"Right": "Down",
		"Down" : "Left",
		"Left" : "Up",
		"Up"   : "Right",
	}
	cur_direction := "Right"
	cur_radius := 1
	cur_radius_time := 0
	cells_encountered := [][]int{
		{rStart, cStart},
	}
	row, col := rStart, cStart

	for {
		for range cur_radius {
			switch cur_direction {
			case "Right": col++
			case "Down" : row++
			case "Left" : col--
			case "Up"   : row--
			}

			if row >= 0 && row < rows && col >= 0 && col < cols {
				cells_encountered = append(cells_encountered, []int{row, col})
			}
		}

		if len(cells_encountered) == cell_count {
			break
		}

		cur_direction = next_direction[cur_direction]
		cur_radius_time++
		if cur_radius_time == 2 {
			cur_radius_time = 0
			cur_radius++
		}
	}

	return cells_encountered
}

File: LeetCode_ProblemSet/Problem885.py
class Solution(object):
    def spiralMatrixIII(self, rows, cols, rStart, cStart):
        """
        :type rows: int
        :type cols: int
        :type rStart: int
        :type cStart: int
        :rtype: List[List[int]]
        """
        cell_count = rows * cols
        next_direction = {
            "Right": "Down",
            "Down" : "Left",
            "Left" : "Up",
            "Up"   : "Right"
        }
        cur_direction = "Right"
        cur_radius = 1
        cur_radius_time = 0
        cells_encountered = [[rStart, cStart]]
        row, col = rStart, cStart
        
        while True:
            for _ in range(cur_radius):
                match cur_direction:
                    case "Right": col += 1
                    case "Down" : row += 1
                    case "Left" : col -= 1
                    case "Up"   : row -= 1
                    
                if 0 <= row < rows and 0 <= col < cols:
                    cells_encountered.append([row, col])
                    
                
            if len(cells_encountered) == cell_count:
                break
            
            cur_direction = next_direction[cur_direction]   
            cur_radius_time += 1
            if cur_radius_time == 2:
                cur_radius_time = 0
                cur_radius += 1
                        
        return cells_encountered